File: erlang/hamming/src/hamming.erl
-module(hamming).

-export([distance/2]).

distance(S1, S2) when length(S1) =/= length(S2) ->
    {error,
     "left and right strands must be of equal "
     "length"};
distance(S1, S2) -> distance(S1, S2, 0).

distance([], [], Diff) -> Diff;
distance([H | R1], [H | R2], Diff) ->
    distance(R1, R2, Diff);
distance([_ | R1], [_ | R2], Diff) ->
    distance(R1, R2, Diff + 1).

File: elixir/hamming/lib/hamming.ex
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance('AAGTCATA', 'TAGCGATC')
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: {:ok, non_neg_integer} | {:error, String.t()}
  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2),
    do: {:error, "Lists must be the same length"}

  def hamming_distance(strand1, strand2), do: do_hamming_distance(Enum.zip(strand1, strand2), 0)

  defp do_hamming_distance([], diff), do: {:ok, diff}

  defp do_hamming_distance([{a, a} | rest], diff),
    do: do_hamming_distance(rest, diff)

  defp do_hamming_distance([_ | rest], diff),
    do: do_hamming_distance(rest, diff + 1)
end
